Converting LeafVersion into an enum

dr-orlovsky opened the pull request:
The original `LeafVersion` implementation was just a newtype around `u8`. I think that having enum explicitly listing consensus script implementation rules may be more beneficial in terms of both code readibility and future use of multiple script types, where `LeafVersion` may operate as a context object provided to `Script` to specify interpretation rules for particular op codes.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -772,23 +776,42 @@ impl LeafVersion {
+        match version {
+            TAPROOT_LEAF_TAPSCRIPT => Ok(LeafVersion::TapScript),
+            TAPROOT_ANNEX_PREFIX => Err(TaprootError::InvalidTaprootLeafVersion(TAPROOT_ANNEX_PREFIX)),
+            even if even & TAPROOT_LEAF_MASK != even => Err(TaprootError::InvalidTaprootLeafVersion(even)),
```
nit: perhaps you want to name this `odd`?

tcharding replied:
Note that the rustdoc says 'even' as well.

sanket1729 replied:
Nice catch, the doc-comment should have said `odd`

Kixunil replied:
Also the parameter name needs to be updated in the docs.

sanket1729 replied:
@dr-orlovsky, can you please address this. I am good with the rest of the PR. 
- Update the docs here (change to odd instead of even)
- The code variable name should be odd instead of even

dr-orlovsky replied:
Thank you for spotting! Addressed with the new force-pushed version.

sanket1729 reviewed (commented):
concept ACK. I really like `from_u8` over `from_byte`, but waiting to see what others think of it.

dr-orlovsky commented:
I actually agree to switch back into `*_u8` since it is not a byte in fact: it will be modified with parity flag

Kixunil commented on src/util/taproot.rs:
```diff
@@ -772,23 +776,42 @@ impl LeafVersion {
+            (LeafVersion::TapScript, true) => fmt::Display::fmt(&TAPROOT_LEAF_TAPSCRIPT, f),
+            (LeafVersion::Future(version), false) => {
+                f.write_str("future(")?;
+                fmt::Display::fmt(version, f)?;
```
I don't think formatter parameters should affect version number. E.g. `"{:10}"` would display `future(.........1)` (spaces instead of dots, GH is broken) instead of `.future(1)`
However, getting parameters right is quite annoying. :(

apoelstra replied:
I think we should just always serialize in hex.

Kixunil replied:
Hmm, version is the kind of thing that I consider naturally decimal.

apoelstra replied:
That's fair.

Ok, "no opinion" on what format it should have :)

dr-orlovsky replied:
I agree with @apoelstra that the script "version", which is not ordinal, is better to be formatted as hex: it will be odd to see the next post-tapscript version as `future(82)` than `future(0x52)`. Decimals will do a bad job here.

I agree with @sanket1729 that we should not use formatting when we use version "number" (actually version id) as a part of other string.

I fixed both of the problems and also changed from `future(x)` to `future_script_0xXX` so it will be more readable. When the version is printed without any other string in alternate representation, I pass the formatting flags to allow users to format it as they like (decimal, prefixed with zeros etc).

Kixunil reviewed (commented):
Concept ACK

apoelstra commented:
concept ACK. I also prefer `_u8` over byte.

dr-orlovsky commented:
@apoelstra @sanket1729 @Kixunil rebased and changed to `*_u8`

apoelstra reviewed (approved):
ACK 1ab928272cdd15f32bc603cd852fcfe3fc6ecb44

I am a little tempted to say that the two error conditions should be separate variants, but I think this is fine.

**Edit:** removed a leading `5` from my hash -- my 5 key and paste-key are right next to each other.

DON-MAC-256 commented:
Concept ACK

In general, we do well to lean further into Rust's excellent enum system.

sanket1729 reviewed (commented):
ACK, but one last thing

RCasatta commented on src/util/taproot.rs:
```diff
@@ -142,6 +143,8 @@ pub const TAPROOT_LEAF_MASK: u8 = 0xfe;
 // https://github.com/bitcoin/bitcoin/blob/e826b22da252e0599c61d21c98ff89f366b3120f/src/script/interpreter.h#L226
 pub const TAPROOT_LEAF_TAPSCRIPT: u8 = 0xc0;
+/// Taproot annex prefix
+pub const TAPROOT_ANNEX_PREFIX: u8 = 0x50;
```
we may want to use this constant also in https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/util/sighash.rs#L728

dr-orlovsky replied:
Did as an additional commit

Kixunil commented:
Note: let's not forget to add this to `non_exhaustive` TODO after it's merged.

dr-orlovsky commented:
Addressed all comments. @Kixunil added to non-exhaustive todo (as a part of your comment https://github.com/rust-bitcoin/rust-bitcoin/issues/510#issuecomment-888989487).

@apoelstra needs re-ACK

Kixunil commented on src/util/taproot.rs:
```diff
@@ -760,16 +764,16 @@ impl ControlBlock {
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-pub struct LeafVersion(u8);
+pub enum LeafVersion {
```
Just realized this changes serde (de)serialization. I think we should still (de)serialize as a single number at least for non-human-friendly encodings.

apoelstra replied:
Good catch! I agree.

Kixunil commented on src/util/taproot.rs:
```diff
@@ -760,16 +764,16 @@ impl ControlBlock {
-        LeafVersion(TAPROOT_LEAF_TAPSCRIPT)
-    }
+    /// Future leaf version
+    Future(u8)
```
This is essentially a public field that allows creating `LeafVersion::Future(TAPROOT_LEAF_TAPSCRIPT)` which could cause breakages/confusion.

I suggest:

```rust
#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
pub struct FutureLeafVersion(u8);

// IMPORTANT: NO PUBLIC CONSTRUCTOR!
// The only way to construct this is by converting u8 to LeafVersion and then extracting it

impl FutureLeafVersion> {
    pub fn to_consensus(self) -> u8 {
        self.0
    }
}

// How to represent it though?
impl Display for FutureLeafVersion { ... }
```

Maybe also perform conversion from consensus encoding to cardinal representation so that we can use `NonZeroUsize`

Kixunil commented on src/util/taproot.rs:
```diff
@@ -780,23 +784,38 @@ impl LeafVersion {
-            Ok(LeafVersion(ver))
-        } else {
-            Err(TaprootError::InvalidTaprootLeafVersion(ver))
+    pub fn from_u8(version: u8) -> Result<Self, TaprootError> {
```
I think this should be called `from_consensus` or something like that to make it clear it doesn't mean 0 => Tapscript.

apoelstra replied:
`from_consensus_u8`?;)

Kixunil replied:
Not sure which is better. As long as it's obvious what is meant by it I'm fine.

dr-orlovsky replied:
@apoelstra I think that `from_consensus_u*` makes sense only when there are multiple representation of the same thing,  like sig hash with `u8` and `u32`

apoelstra replied:
Makes sense. I'm happy with whatever name y'all want.

Kixunil reviewed (changes requested):
I think serialization issue is serious enough that I'd like at least some good argument why keep it.
I'd also really, really like to avoid weird states by using `FutureLeafVersion`.

The potential confusion related to `from_u8` is not great although not strictly related to this PR, so willing to accept in followup PR.

dr-orlovsky commented:
@Kixunil I did as you have suggested. Interesting enough this breaks unit tests, since they include deserialization of an invalid leaf version. I assume this must not error and deserialize correctly, otherwise taproot would be a hard fork and not a soft fork. Basically, we need to allow processing (but not constructing?) invalid control blocks.

I propose to add third variant (`LeafVersion::Invalid(InvalidLeafVersion))`) which will be produced only during deserialization and can´t be constructed manually.  @apoelstra what do you think?

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -757,20 +761,68 @@ impl ControlBlock {
+    }
+}
+
+impl fmt::Display for FutureLeafVersion {
```
@Kixunil I think this (unlike `LeafVersion`) should be allowed to be displayed as a simple number with whatever formatting options were used by the caller.

This is an internal type which can be obtained only from `match`ing over `LeafVersion` and thus will be rarely print directly.

Kixunil replied:
Sure, I don't feel strongly about the representation. But maybe also implement hex traits?

Kixunil replied:
TODO: create good first issue for hex traits

dr-orlovsky replied:
I clearly remember that I was writing one here. Probably commit got lost through multiple rebases...

Kixunil commented:
I don't understand why rejecting an invalid control block would be hard fork and not soft fork. Anyway, this is outside of my knowledge. If accepting invalid versions is truly needed then I'm not against having another variant/type but not currently entirely sure it's the best way.

Kixunil commented on src/util/taproot.rs:
```diff
@@ -780,23 +832,61 @@ impl LeafVersion {
-        self.0
+        self.into_consensus()
+    }
+}
```
Same (or even worse) as above.

Kixunil reviewed (commented):
`impl Into<u8>` doesn't seem right, so not ACKing yet.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -757,20 +761,68 @@ impl ControlBlock {
+    }
+}
+
+impl Into<u8> for FutureLeafVersion {
```
Agree. I was not sure about it and kept to simplify serde impl. But will remove it and do manual serialization

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -780,23 +832,61 @@ impl LeafVersion {
+                where
+                    E: serde::de::Error,
+            {
+                LeafVersion::from_consensus(value).map_err(|err| E::custom(err))
```
`E::unexpected_value` requires to list all the expected values, which makes this looking ugly

Kixunil replied:
Not really, you can describe it with a string `"consensus-encoded leaf version as u8"` or something similar

apoelstra commented:
If our tests assume you can deserialize invalid control blocks then the tests are wrong. There is no reason to support this.

And no, disallowing invalid data does not make Taproot a hardfork.

dr-orlovsky commented:
@apoelstra from what I understand, pre-taproot block may contain witness v1 output spent by anybody with arbitrary script (before taproot activation) including use of invalid control block.

The test which is failing is taken from bitcoin core, so probably we should double-check what it is... I will try to figure this out

Kixunil commented:
@dr-orlovsky maybe the shouldn't be parsed as control blocks pre-taproot activation?

apoelstra commented:
We don't need to support pre-Taproot Taproot-like spends in this library.

Kixunil commented:
From README:

> This library **must not** be used for consensus code

and

> patches to fix specific consensus incompatibilities are welcome.

Suggest that it's up to PR author to decide.

IMO it'd be faster to just skip it now, releasing Taproot is more important.
If someone wants to do it (later), I'd prefer using some flag controlling whether Taproot is activated and avoiding parsing control blocks completely if it isn't. Similarly to how Core does it.

dr-orlovsky commented:
Ok, rebased & removed failed tests with invalid control blocks. Ready for re-review

Kixunil commented on src/util/taproot.rs:
```diff
@@ -779,23 +837,87 @@ impl LeafVersion {
+            (LeafVersion::TapScript, true) => fmt::Display::fmt(&TAPROOT_LEAF_TAPSCRIPT, f),
+            (LeafVersion::Future(version), false) => write!(f, "future_script_{:#02x}", version.0),
+            (LeafVersion::Future(version), true) => fmt::Display::fmt(version, f),
+        }
```
Very nice that you implement alternate formatting, however the convention is that alternate true means more human-friendly/verbose representation, so it appears it's inverted here.

Kixunil replied:
Don't feel like blocking because of this but we should try getting this in before release if we have time.

Kixunil commented on src/util/taproot.rs:
```diff
@@ -779,23 +837,87 @@ impl LeafVersion {
+        match (self, f.alternate()) {
+            (LeafVersion::TapScript, false) => f.write_str("tapscript"),
+            (LeafVersion::TapScript, true) => fmt::Display::fmt(&TAPROOT_LEAF_TAPSCRIPT, f),
+            (LeafVersion::Future(version), false) => write!(f, "future_script_{:#02x}", version.0),
```
Super-unimortant: `.0` could be dropped here.

dr-orlovsky replied:
Better not to: we may change the way FutureLeafVersion is displayed in a future and this will break formatting here

Kixunil commented on src/util/taproot.rs:
```diff
@@ -779,23 +837,87 @@ impl LeafVersion {
+    }
+}
+
+impl fmt::UpperHex for LeafVersion {
```
Comment on these (consensus encoding) may also be helpful.

Kixunil commented on src/util/taproot.rs:
```diff
@@ -779,23 +837,87 @@ impl LeafVersion {
+
+#[cfg(feature = "serde")]
+#[cfg_attr(docsrs, doc(cfg(feature = "serde")))]
+impl<'de> ::serde::Deserialize<'de> for LeafVersion {
```
Shame you forgot about doc comment. Not critical.

Kixunil replied:
TODO: create good first issue

dr-orlovsky replied:
Hm, which kind of doc comment can be put here?

Kixunil replied:
Something like "(De)serializes `LeafVersion` as `u8` using consensus encoding."

Kixunil reviewed (dismissed):
ACK a3642e8977ac3b19994644bb25c1c48e464f115e

Just non-critical style comments, can be addressed later or ignored.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -790,16 +827,16 @@ impl LeafVersion {
             TAPROOT_ANNEX_PREFIX => Err(TaprootError::InvalidTaprootLeafVersion(TAPROOT_ANNEX_PREFIX)),
-            odd if odd & TAPROOT_LEAF_MASK != odd => Err(TaprootError::InvalidTaprootLeafVersion(odd)),
-            future => Ok(LeafVersion::Future(future)),
+            0xC1 => Err(TaprootError::InvalidTaprootLeafVersion(0xC1)),
```
Any special reason why this needs a special line? I think this would be handled by the `future` match arm in the next line.

dr-orlovsky replied:
Fixed

sanket1729 commented:
> Interestingly enough disallowing invalid leaf versions breaks unit tests, since they include deserialization of invalid control blocks. I assume this must be allowed, otherwise taproot would be a hard fork and not a soft fork. Basically, we need to allow processing (but not constructing?) invalid leaf versions?!

@dr-orlovsky @apoelstra The current implementation is master correct. As I mentioned in the review, I think the futureleafversion is incorrect. I had those test vectors from cross-testing with bitcoin core. 

From BIP 341: 

> In addition, in order to support some forms of static analysis that rely on being able to identify script spends without access to the output being spent, it is **recommended**  ....

There can be an update to use leaf version, 0x02, but BIP341 does not recommend doing it. But an update would be consensus accepted. Thus, we must parse those leaf versions correctly.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -1059,6 +1096,8 @@ mod test {
         let secp  = Secp256k1::verification_only();
         // test vectors obtained from printing values in feature_taproot.py from bitcoin core
+        // excluding four test cases using invalid leaf version in the control block
+        // see details in <https://github.com/rust-bitcoin/rust-bitcoin/pull/718#issuecomment-1007596228>
```
No need to exclude these tests, if you change the above, the tests can then be uncommented.

apoelstra replied:
Regardless they should definitely not be called **invalid** if they are not actually invalid.

Sorry for contributing to confusion instead of just checking the BIP myself.

dr-orlovsky replied:
@apoelstra sorry, now I do not understand at all what should be done with these tests or `LeafVersion` type.

apoelstra replied:
@dr-orlovsky the tests should be restored. We should only reject odd values or `0x50`, as the BIP says in footnote 7. Nothing else should be rejected.

dr-orlovsky replied:
How to classify all values which are not `0xC0`, `LeafVersion::Future`?

apoelstra replied:
Everything that isn't known or invalid is `Future` (even though the real-life future will not include BIPs that ignore the advice in BIP 341).

dr-orlovsky replied:
done

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -760,14 +760,51 @@ impl ControlBlock {
+            0xE0 | 0xE2 | 0xE4 | 0xE6 | 0xE8 | 0xEA | 0xEC | 0xEE |
+            0xF0 | 0xF2 | 0xF4 | 0xF6 | 0xF8 | 0xFA | 0xFC | 0xFE |
+            // and also 0x66, 0x7e, 0x80, 0x84, 0x96, 0x98, 0xba, 0xbc, 0xbe
+            0x66 | 0x7E | 0x80 | 0x84 | 0x96 | 0x98 | 0xBA | 0xBC | 0xBE => Ok(FutureLeafVersion(version)),
```
Fixed

Kixunil commented on src/util/taproot.rs:
```diff
@@ -756,20 +760,54 @@ impl ControlBlock {
+/// Inner type representing future (non-tapscript) leaf versions. See [`LeafVersion::Future`].
+///
+/// NB: NO PUBLIC CONSTRUCTOR!
+/// The only way to construct this is by converting u8 to LeafVersion and then extracting it.
```
TODO: create good first issue for rewriting to `` `u8` `` and ``[`LeafVersion`]``

dr-orlovsky replied:
Sorry, I am not following what do you mean here

Kixunil replied:
Putting backticks and square brackets around them :)

Kixunil reviewed (approved):
Even though I'm not happy about inverted `alternate()` proceeding is better.

ACK ef8a3a839e5f0faccf9f6299b295309f2eab3727

sanket1729 reviewed (approved):
crACK ef8a3a839e5f0faccf9f6299b295309f2eab3727. Waiting a day to let others complete review before merging.

apoelstra commented on src/util/taproot.rs:
```diff
@@ -790,16 +821,15 @@ impl LeafVersion {
             TAPROOT_ANNEX_PREFIX => Err(TaprootError::InvalidTaprootLeafVersion(TAPROOT_ANNEX_PREFIX)),
-            odd if odd & TAPROOT_LEAF_MASK != odd => Err(TaprootError::InvalidTaprootLeafVersion(odd)),
-            future => Ok(LeafVersion::Future(future)),
+            future => FutureLeafVersion::from_consensus(future).map(LeafVersion::Future),
```
In ef8a3a839e5f0faccf9f6299b295309f2eab3727:

We should refuse to parse odd leaf versions.

dr-orlovsky replied:
Thry are refused at the level of `FutureLeafVersion`

apoelstra replied:
Derp, you're right, thanks.

apoelstra reviewed (approved):
ACK ef8a3a839e5f0faccf9f6299b295309f2eab3727

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -756,20 +760,74 @@ impl ControlBlock {
+    /// # Errors
+    /// - If the last bit of the `version` is odd.
+    /// - If the `version` is 0x50 ([`TAPROOT_ANNEX_PREFIX`]).
+    /// - If the `version` is not a valid [`LeafVersion`] byte.
```
Not nice anymore: we do not error on invalid LeafVersion bytes other than annex. Will provide a fix for this doc

dr-orlovsky commented:
Addresses all your addressable requests in #761
